_posts/2019-07-27-GettingStartedWithProtobufGrpc.md

```diff
@@ -0,0 +1,306 @@
+- 2. **Efficient serialization**: `Protobuf` is a language-neutral way for serializing structured data. They are also very efficient and packed, while JSON, on the other hand, is textual inherently. If you have worked with JSON and XML, Protobuf lies in the same space with an added advantage that protobuf compiler can translate your data definitions to the language of your choice. Therefore, you create your structure once and use it across all supported languages.
+- 3. **RPC style communication**: `REST` encourages strict CRUD operations on a resource, however with `RPC` server can provide more general computation functions like `generateReport()`.
+
+
+References:
+- [http2-vs-http1](https://imagekit.io/blog/http2-vs-http1-performance/)
+- [comparing-serilization-formats](http://labs.criteo.com/2017/05/serialization/)
+- [rest-vs-rpc-style](https://apihandyman.io/do-you-really-know-why-you-prefer-rest-over-rpc/)
+
+### 1.1. Features.
+
+Apart from providing a robust framework for client-server communication, gRPC provides a few other features which make it an indispensable choice for modern application development.
+
+- Strongly typed payload and service definition. Ensures consistent contract between the client and server.
+- API evolution, versioning, and backward compatibility. gRPC APIs can be changed transparently without breaking client contracts.
+- Supports unary and streaming communication with bidirectional streaming support.
+- Pluggable auth, tracing, load balancing and health checking.
+- Code generation for multiple programming languages.
+- More performant than traditional REST+JSON.
+- Supports reliable gRPC clients as well as clients for web, Android, and iOS.
+- Cloud native and integrates well with cloud technologies.
+
+### 1.2. Supported programming languages.
+Below are the officially supported programming languages:
+- C++
+- Java (including support for Android)
+- Object-C (for iOS)
+- Python
+- Ruby
+- Go
+- C#
+- Node.js
+- PHP
+- Dart (beta)
+
+Other third-party supported languages:
+- Haskell
+- Erlang
+- Scala
+- Elixir
+- Rust
+- Elm
+- TypeScript
+
+### 1.3. Supported data formats. 
+
+By default, gRPC uses protocol buffers, Google’s mature open source mechanism for serializing structured data. However, since protocol buffers act as serialization layer, it is possible to replace it with other content types or provide your own implementations.
+
+From gRPC website:
+
+> gRPC is designed to be extensible to support multiple content types. The initial release contains support for Protobuf and with external support for other content types such as FlatBuffers and Thrift, at varying levels of maturity.
+
+References:
+- [gRPC-with-flatbuffers](https://grpc.io/blog/flatbuffers)
+### 1.4. Typical applications that can be built with gRPC.
```

Comment:
nit: titles could be separated from paragraphs to make text file more readable.

Reply:
Fixed